perception/map_based_prediction/include/map_based_prediction_node.hpp

```diff
@@ -0,0 +1,200 @@
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+namespace map_based_prediction
+{
+
+struct ObjectData
+{
+  std_msgs::msg::Header header;
+  lanelet::ConstLanelets current_lanelets;
+  lanelet::ConstLanelets future_possible_lanelets;
+  geometry_msgs::msg::Pose pose;
+  geometry_msgs::msg::Twist twist;
+  double time_delay;
+};
+
+enum class Maneuver {
+  LANE_FOLLOW = 0,
+  LEFT_LANE_CHANGE = 1,
+  RIGHT_LANE_CHANGE = 2,
+};
+
+struct LaneletData
+{
+  lanelet::Lanelet lanelet;
+  float probability;
+};
+
+struct PredictedRefPath
+{
+  float probability;
+  PosePath path;
+  Maneuver maneuver;
+};
+
+using LaneletsData = std::vector<LaneletData>;
+using ManeuverProbability = std::unordered_map<Maneuver, float>;
+using autoware_auto_mapping_msgs::msg::HADMapBin;
+using autoware_auto_perception_msgs::msg::ObjectClassification;
+using autoware_auto_perception_msgs::msg::PredictedObject;
+using autoware_auto_perception_msgs::msg::PredictedObjectKinematics;
+using autoware_auto_perception_msgs::msg::PredictedObjects;
+using autoware_auto_perception_msgs::msg::PredictedPath;
+using autoware_auto_perception_msgs::msg::TrackedObject;
+using autoware_auto_perception_msgs::msg::TrackedObjectKinematics;
+using autoware_auto_perception_msgs::msg::TrackedObjects;
+
+class MapBasedPredictionNode : public rclcpp::Node
+{
+public:
+  explicit MapBasedPredictionNode(const rclcpp::NodeOptions & node_options);
+
+private:
+  // ROS Publisher and Subscriber
+  rclcpp::Publisher<PredictedObjects>::SharedPtr pub_objects_;
+  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr pub_debug_markers_;
+  rclcpp::Subscription<TrackedObjects>::SharedPtr sub_objects_;
+  rclcpp::Subscription<HADMapBin>::SharedPtr sub_map_;
+
+  // Object History
+  std::unordered_map<std::string, std::deque<ObjectData>> objects_history_;
+
+  // Lanelet Map Pointers
+  std::shared_ptr<lanelet::LaneletMap> lanelet_map_ptr_;
+  std::shared_ptr<lanelet::routing::RoutingGraph> routing_graph_ptr_;
+  std::shared_ptr<lanelet::traffic_rules::TrafficRules> traffic_rules_ptr_;
+
+  // Pose Transform Listener
+  tier4_autoware_utils::TransformListener transform_listener_{this};
+
+  // Path Generator
+  std::shared_ptr<PathGenerator> path_generator_;
+
+  // Parameters
+  bool enable_delay_compensation_;
+  double prediction_time_horizon_;
+  double prediction_sampling_time_interval_;
+  double min_velocity_for_map_based_prediction_;
+  double debug_accumulated_time_;
+  double dist_threshold_for_searching_lanelet_;
+  double delta_yaw_threshold_for_searching_lanelet_;
+  double sigma_lateral_offset_;
+  double sigma_yaw_angle_;
+  double object_buffer_time_length_;
+  double history_time_length_;
+  double dist_ratio_threshold_to_left_bound_;
+  double dist_ratio_threshold_to_right_bound_;
+  double diff_dist_threshold_to_left_bound_;
+  double diff_dist_threshold_to_right_bound_;
+  double reference_path_resolution_;
```

Comment:
please use floats where double precision is not required.

Reply:
I think these variables should use double precision

Reply:
I don't think it's necessary.
What is the actual number to be substituted?